R/helpers.R

```diff
@@ -698,10 +678,191 @@ h_is_positive_definite <- function(x, tolerance = 1e-08) {
+#'   JAGS code into a file.
+#'   Due to technical issues related to conversion of numbers to scientific
+#'   format, it is required that the body of a model function does not contain
+#'   `TEMP_NUM_PREF_` or `_TEMP_NUM_SUF` character constants in its body.
+#'
+#' @param model (`function`)\cr function containing the JAGS model.
+#' @param file (`string`)\cr the name of the file where the model will be saved.
+#' @param digits (`count`)\cr a desired number of significant digits for
+#'   for numbers used in JAGS input, see [`formatC`].
+#' @return Nothing, but as a side effect, the model file is written.
+#'
+#' @example examples/helpers-write_model.R
+#' @export
+#'
+h_write_model <- function(model, file = "model.jags", digits = 5) {
+  assert_function(model)
+  assert_string(file)
+  assert_count(digits)
+
+  # Replace scientific notation.
+  model_sci_replaced <- h_rapply(
+    x = body(model),
+    fun = h_format_number,
+    classes = c("integer", "numeric"),
+    digits = digits,
+    prefix = "TEMP_NUM_PREF_",
+    suffix = "_TEMP_NUM_SUF"
+  )
+  # Transform `model` body into character vector.
+  model_text <- deparse(model_sci_replaced, control = NULL)
+  model_text <- gsub("\"TEMP_NUM_PREF_|_TEMP_NUM_SUF\"", "", model_text)
+  model_text <- gsub("%_% ", "", model_text)
+  model_text <- c("model", model_text)
+
+  writeLines(model_text, con = file)
+}
+
+#' Recursively Apply a Function to a List
+#'
+#' @description `r lifecycle::badge("experimental")`
+#'
+#' This helper function is conceptually quite similar to [`base::rapply`]
+#' function with two major differences (see the note for more details).
+#' It recursively iterates through a "list-like" object and it checks whether
+#' an element is of a given class. If it so, then it replaces that element by
+#' the result of an execution of a given function. Otherwise, and if the element
+#' is of length greater than 1 (i.e. not a scalar), it replaces that element by
+#' the result of `h_rapply`, recursively called for that element. In the
+#' remaining case, that is, the element is not of a given class and is a scalar,
+#' then that element remains unchanged.
+#'
+#' @note This helper function is conceptually similar the same as `rapply`
+#'   function. However, it differs from `rapply` in two major ways. First, the
+#'   `h_rapply` is not limited to objects of type `list` or `expression` only.
+#'   It can be any "list-like" object of any type for which `[[` operator is
+#'   defined. This can be, for example, an object of type `language`, often
+#'   obtained from the `body` function. The second difference is that the
+#'   flexibility of `rapply` on `how` the result is structured is not available
+#'   with `h_rapply` for the user. That is, with `h_rapply` each element of `x`,
+#'   which has a class included in `classes`, is replaced by the result of
+#'   applying `fun` to the element.
+#'
+#' @param x \cr any "list-like" object for which `[[` operator is defined.
```

Comment:
```suggestion
#' @param x any "list-like" object for which `[[` operator is defined.
```
(if we don't give class in parenthesis then no need to have line break)